radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,701 @@
+        self.revisions().next_back()
+    }
+}
+
+impl store::FromHistory for Proposal {
+    type Action = Action;
+    type Error = ApplyError;
+
+    fn type_name() -> &'static TypeName {
+        &*TYPENAME
+    }
+
+    fn apply(&mut self, ops: impl IntoIterator<Item = Op>) -> Result<(), Self::Error> {
+        for op in ops {
+            let id = op.id();
+            let author = Author::new(op.author);
+            let timestamp = op.timestamp;
+
+            match op.action {
+                Action::Accept {
+                    revision,
+                    signature,
+                } => match self.revisions.get_mut(&revision) {
+                    Some(Redactable::Present(revision)) => {
+                        revision.accept(op.author, signature, op.clock)
+                    }
+                    Some(Redactable::Redacted) => return Err(ApplyError::Redacted(revision)),
+                    None => return Err(ApplyError::Missing(revision)),
+                },
+                Action::Close => self.state.set(State::Closed, op.clock),
+                Action::Edit { title, description } => {
+                    self.title.set(title, op.clock);
+                    self.description.set(description, op.clock);
+                }
+                Action::Publish => self.state.set(State::Published, op.clock),
+                Action::Redact { revision } => {
+                    if let Some(revision) = self.revisions.get_mut(&revision) {
+                        revision.merge(Redactable::Redacted);
+                    } else {
+                        return Err(ApplyError::Missing(revision));
+                    }
+                }
+                Action::Reject { revision } => match self.revisions.get_mut(&revision) {
+                    Some(Redactable::Present(revision)) => revision.reject(op.author, op.clock),
+                    Some(Redactable::Redacted) => return Err(ApplyError::Redacted(revision)),
+                    None => return Err(ApplyError::Missing(revision)),
+                },
+                Action::Revision { proposed } => self.revisions.insert(
+                    id,
+                    Redactable::Present(Revision::new(author, proposed, timestamp)),
+                ),
+                Action::Thread { revision, action } => match self.revisions.get_mut(&revision) {
+                    Some(Redactable::Present(revision)) => revision
+                        .discussion
+                        .apply([cob::Op::new(action, op.author, op.timestamp, op.clock)])?,
+                    Some(Redactable::Redacted) => return Err(ApplyError::Redacted(revision)),
+                    None => return Err(ApplyError::Missing(revision)),
+                },
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Verdict {
+    /// An accepting verdict must supply the [`Signature`] over the
+    /// new proposed [`Doc`].
+    Accept(Signature),
```

Comment:
Are we still ok with signing the whole doc, vs. the blob id? It's still time to change that.

Reply:
iirc, signing the blob id made sense for trying to match the `Patch` cob, but that turned out to be tricky. The only pro I see for signing the doc is that it's stable in the face of the SHA scheme changing for Git. Otherwise, they seem equivalent.

wdyt?

Reply:
Hmm yeah, I guess signing hashes is much more standard generally, because you can verify the signature with less data and the signing is also faster. I actually think we should maybe move to signing the `git hash-object` hash of the blob (in a different pr though).

Reply:
Sounds good :)